dipy/reconst/fwdti.py

```diff
@@ -0,0 +1,608 @@
+            except KeyError:
+                e_s = '"' + str(fit_method) + '" is not a known fit '
+                e_s += 'method, the fit method should either be a '
+                e_s += 'function or one of the common fit methods'
+                raise ValueError(e_s)
+        self.fit_method = fit_method
+        self.design_matrix = design_matrix(self.gtab)
+        self.args = args
+        self.kwargs = kwargs
+        self.min_signal = self.kwargs.pop('min_signal', None)
+        if self.min_signal is not None and self.min_signal <= 0:
+            e_s = "The `min_signal` key-word argument needs to be strictly"
+            e_s += " positive."
+            raise ValueError(e_s)
+
+    def fit(self, data, mask=None):
+        """ Fit method of the free water elimination DTI model class
+        Parameters
+        ----------
+        data : array
+            The measured signal from one voxel.
+        mask : array
+            A boolean array used to mark the coordinates in the data that
+            should be analyzed that has the shape data.shape[:-1]
+        """
+        if mask is None:
+            # Flatten it to 2D either way:
+            data_in_mask = np.reshape(data, (-1, data.shape[-1]))
+        else:
+            # Check for valid shape of the mask
+            if mask.shape != data.shape[:-1]:
+                raise ValueError("Mask is not the same shape as data.")
+            mask = np.array(mask, dtype=bool, copy=False)
+            data_in_mask = np.reshape(data[mask], (-1, data.shape[-1]))
+
+        if self.min_signal is None:
+            min_signal = _min_positive_signal(data)
+        else:
+            min_signal = self.min_signal
+
+        data_in_mask = np.maximum(data_in_mask, min_signal)
+        params_in_mask = self.fit_method(self.design_matrix, data_in_mask,
+                                         *self.args, **self.kwargs)
+
+        if mask is None:
+            out_shape = data.shape[:-1] + (-1, )
+            fwdti_params = params_in_mask.reshape(out_shape)
+        else:
+            fwdti_params = np.zeros(data.shape[:-1] + (14,))
+            fwdti_params[mask, :] = params_in_mask
+
+        return FreeWaterTensorFit(self, fwdti_params)
+
+    def predict(self, fwdti_params):
+        """
+        Predict a signal for this TensorModel class instance given parameters.
+        Parameters
+        ----------
+        fwdti_params : ndarray
+            The last dimension should have 14 parameters: the 12 tensor
+            parameters (3 eigenvalues, followed by the 3 corresponding
+            eigenvectors), the volume fraction of the free water compartment
+            and the estimate of the non diffusion-weighted signal S0.
+        S0 : float or ndarray
+            The non diffusion-weighted signal in every voxel, or across all
+            voxels. Default: 1
+        """
+        return fwdti_prediction(fwdti_params, self.gtab)
```

Comment:
I think you got the wrong line - step should be used in the function of FreeWaterTensorFit class and not in the function of FreeWaterTensorModel, right?